sync/syncer.go

```diff
@@ -291,149 +399,90 @@ func (s *Syncer) Txs(mb *types.Block) ([]*types.SerializableTransaction, error)
 
 	//map and sort atxs
-	if len(missing) > 0 {
-		output := s.fetchWithFactory(ATxReqFactory(missing), 1)
+	if len(missingInDb) > 0 {
+
+		output := s.fetchWithFactory(NewNeighborhoodWorker(s, 1, ATxReqFactory(missingInDb)))
 		for out := range output {
-			ntxs := out.([]types.ActivationTx)
-			for _, atx := range ntxs {
-				atxMap[atx.Id()] = &atx
+			atxs := out.([]types.ActivationTx)
+			for _, atx := range atxs {
+				if err := s.ValidateAtx(&atx); err != nil {
+					s.Warning("atx %v not valid %v", atx.ShortId(), err)
+					continue
+				}
+				unprocessedAtxs[atx.Id()] = &atx
 			}
 		}
 	}
 
-	if fetchAssociated {
-		if _, ok := atxMap[mb.BlockHeader.ATXID]; !ok {
-			return nil, nil, errors.New(fmt.Sprintf("could not fetch associated %v", hex.EncodeToString(mb.BlockHeader.ATXID.Bytes())))
-		}
-	}
-
-	atxs = make([]*types.ActivationTx, 0, len(mb.AtxIds))
-	for _, t := range mb.AtxIds {
-		if tx, ok := atxMap[t]; ok {
-			atxs = append(atxs, tx)
-		} else if tx, ok := dbAtxs[t]; ok {
+	atxs = make([]*types.ActivationTx, 0, len(atxIds))
+	for _, id := range atxIds {
+		if tx, ok := unprocessedAtxs[id]; ok {
 			atxs = append(atxs, tx)
-		} else {
-			return nil, nil, errors.New(fmt.Sprintf("could not fetch atx %v", hex.EncodeToString(t.Bytes())))
-		}
-	}
-	return atxs, associated, nil
-}
-
-func (s *Syncer) fetchLayerBlockIds(m map[string]p2p.Peer, lyr types.LayerID) ([]types.BlockID, error) {
-	//send request to different users according to returned hashes
-	v := make([]p2p.Peer, 0, len(m))
-	for _, value := range m {
-		v = append(v, value)
-	}
-
-	wrk, output := NewPeersWorker(s, v, &sync.Once{}, LayerIdsReqFactory(lyr))
-	go wrk.Work()
-
-	out := <-output
-	if out == nil {
-		return nil, errors.New("could not get layer ids from any peer")
-	}
-
-	idSet := make(map[types.BlockID]struct{}, s.LayerSize)
-	ids := make([]types.BlockID, 0, s.LayerSize)
-
-	//filter double ids
-	for _, bid := range out.([]types.BlockID) {
-		if _, exists := idSet[bid]; !exists {
-			idSet[bid] = struct{}{}
-			ids = append(ids, bid)
-		}
-	}
-
-	return ids, nil
-}
-
-type peerHashPair struct {
-	peer p2p.Peer
-	hash []byte
-}
-
-func (s *Syncer) fetchLayerHashes(lyr types.LayerID) (map[string]p2p.Peer, error) {
-	// get layer hash from each peer
-	wrk, output := NewPeersWorker(s, s.GetPeers(), &sync.Once{}, HashReqFactory(lyr))
-	go wrk.Work()
-	m := make(map[string]p2p.Peer)
-	for out := range output {
-		pair := out.(*peerHashPair)
-		if pair == nil { //do nothing on close channel
+		} else if _, ok := dbAtxs[id]; ok {
 			continue
+		} else {
+			return nil, errors.New(fmt.Sprintf("could not fetch atx %v", id.ShortId()))
 		}
-		m[string(pair.hash)] = pair.peer
 	}
-	if len(m) == 0 {
-		return nil, errors.New("could not get layer hashes from any peer")
-	}
-	return m, nil
+	return atxs, nil
 }
 
-func (s *Syncer) fetchWithFactory(refac RequestFactory, workers int) chan interface{} {
+func (s *Syncer) fetchWithFactory(wrk worker) chan interface{} {
```

Comment:
please add a comment on this method and any related since it is not trivial at all